### PR TITLE
Bump pius to 3.0.0 and add Nicholas Meyer as proxy maintainer

### PR DESCRIPTION
**Pull request by aristocrates**
* https://github.com/jaymzh/pius/releases/tag/v3.0.0
* Migrates to python 3
* Drops GnuPG 1 support

You can confirm that @jaymzh's signature is valid for the tarball and that the b2sum and sha512sum match what is in my commit:
* https://github.com/jaymzh/pius/releases/download/v3.0.0/pius-3.0.0.tar.bz2
* https://github.com/jaymzh/pius/releases/download/v3.0.0/pius-3.0.0.tar.bz2.asc

Bug: https://bugs.gentoo.org/702762
Closes: https://bugs.gentoo.org/702762
Signed-off-by: Nicholas Meyer <nickaristocrates@gmail.com>

### CODE REVIEW

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+PYTHON_COMPAT=( python3_6 python3_7 python3_8 )
```
Comment (jaymzh) [inline app-crypt/pius/pius-3.0.0.ebuild]: 3.7 or later is required for 3.0.0

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# Copyright 1999-2018 Gentoo Foundation
```
Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild]: Update this copyright header, `-2020 Gentoo Authors`. You can commit with a tool called **repoman** which will fix most of these common mistakes for you, and even sign-off your commit automatically once you set it up.

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+inherit distutils-r1
+
+DESCRIPTION="A tool for signing and email all UIDs on a set of PGP keys"
+HOMEPAGE="https://www.phildev.net/pius/"
+SRC_URI="https://github.com/jaymzh/pius/releases/download/v${PV}/pius-${PV}.tar.bz2"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE=""
```
Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild]: Empty variable can be removed for a cleaner ebuild.

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+PYTHON_COMPAT=( python3_7 python3_8 )
```
Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild]: One empty line below EAPI please for tree consistency.

Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild]: You can use `PYTHON_COMPAT=( python3_{7,8} )`

**Comment by juippis** [comment]
Oh I noted upstream seems to have a file to test python runtime, could you try implementing it?

**Comment by aristocrates** [comment]
@juippis With `app-crypt/pius[test]` and `FEATURES=test`, 3.0.0 passes the test suite with this message:

```
>>> Test phase: app-crypt/pius-3.0.0
test_convert_from_v1 (test_state.TestFileConversion) ... ok
test_convert_from_v2 (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_new_data (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_no_new_data (test_state.TestFileConversion) ... ok
test_read_v1_writes_v3 (test_state.TestFileConversion) ... ok
test_read_v2_writes_v3 (test_state.TestFileConversion) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.015s

OK
```

and if you break the test suite and try to merge, you get a message like this:

```
>>> Test phase: app-crypt/pius-3.0.1
test_convert_from_v1 (test_state.TestFileConversion) ... FAIL
test_convert_from_v2 (test_state.TestFileConversion) ... FAIL
test_read_empty_writes_v3_with_new_data (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_no_new_data (test_state.TestFileConversion) ... ok
test_read_v1_writes_v3 (test_state.TestFileConversion) ... FAIL
test_read_v2_writes_v3 (test_state.TestFileConversion) ... FAIL

======================================================================
FAIL: test_convert_from_v1 (test_state.TestFileConversion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/portage/app-crypt/pius-3.0.1/work/pius-3.0.1/test/libpius/test_state.py", line 24, in test_convert_from_v1
    self.kSAMPLE,
AssertionError: {'one': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'two': {'O[82 chars]one}} != {'one': {'OUTBOUNDD': 'SIGNED', 'INBOUND': None}, 'two': {'[83 chars]one}}
- {'one': {'INBOUND': None, 'OUTBOUND': 'SIGNED'},
+ {'one': {'INBOUND': None, 'OUTBOUNDD': 'SIGNED'},
?                                    +

   'three': {'INBOUND': None, 'OUTBOUND': 'SIGNED'},
   'two': {'INBOUND': None, 'OUTBOUND': 'SIGNED'}}

======================================================================
FAIL: test_convert_from_v2 (test_state.TestFileConversion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/var/tmp/portage/app-crypt/pius-3.0.1/work/pius-3.0.1/test/libpius/test_state.py", line 34, in test_convert_from_v2
    self.kSAMPLE,
AssertionError: {'one': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'two': {'O[82 chars]one}} != {'one': {'OUTBOUNDD': 'SIGNED', 'INBOUND': None}, 'two': {'[83 chars]one}}
- {'one': {'INBOUND': None, 'OUTBOUND': 'SIGNED'},
+ {'one': {'INBOUND': None, 'OUTBOUNDD': 'SIGNED'},
?                                    +

   'three': {'INBOUND': None, 'OUTBOUND': 'SIGNED'},
   'two': {'INBOUND': None, 'OUTBOUND': 'SIGNED'}}

======================================================================
FAIL: test_read_v1_writes_v3 (test_state.TestFileConversion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/mock.py", line 1255, in patched
    return func(*args, **keywargs)
  File "/var/tmp/portage/app-crypt/pius-3.0.1/work/pius-3.0.1/test/libpius/test_state.py", line 47, in test_read_v1_writes_v3
    write_file.assert_called_once_with(temp)
  File "/usr/lib/python3.7/unittest/mock.py", line 884, in assert_called_once_with
    return self.assert_called_with(*args, **kwargs)
  File "/usr/lib/python3.7/unittest/mock.py", line 873, in assert_called_with
    raise AssertionError(_error_message()) from cause
AssertionError: Expected call: write_file({'one': {'OUTBOUNDD': 'SIGNED', 'INBOUND': None}, 'two': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'three': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'four': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, '_meta': {'version': 3}})
Actual call: write_file({'one': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'two': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'three': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'four': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, '_meta': {'version': 3}})

======================================================================
FAIL: test_read_v2_writes_v3 (test_state.TestFileConversion)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/mock.py", line 1255, in patched
    return func(*args, **keywargs)
  File "/var/tmp/portage/app-crypt/pius-3.0.1/work/pius-3.0.1/test/libpius/test_state.py", line 59, in test_read_v2_writes_v3
    write_file.assert_called_once_with(temp)
  File "/usr/lib/python3.7/unittest/mock.py", line 884, in assert_called_once_with
    return self.assert_called_with(*args, **kwargs)
  File "/usr/lib/python3.7/unittest/mock.py", line 873, in assert_called_with
    raise AssertionError(_error_message()) from cause
AssertionError: Expected call: write_file({'one': {'OUTBOUNDD': 'SIGNED', 'INBOUND': None}, 'two': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'three': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'four': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, '_meta': {'version': 3}})
Actual call: write_file({'one': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'two': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'three': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, 'four': {'OUTBOUND': 'SIGNED', 'INBOUND': None}, '_meta': {'version': 3}})

----------------------------------------------------------------------
Ran 6 tests in 0.018s

FAILED (failures=4)
 * ERROR: app-crypt/pius-3.0.1::nmeyer failed (test phase):
 *   Tests failed
 * 
 * Call stack:
 *     ebuild.sh, line 125:  Called src_test
 *   environment, line 2874:  Called die
 * The specific snippet of code:
 *       "${S}"/run_tests || die "Tests failed"
```

The diff I used to test breaking the pius test suite:

```diff
diff --git a/test/libpius/test_state.py b/test/libpius/test_state.py
index 7af1891..1bc6ca7 100755
--- a/test/libpius/test_state.py
+++ b/test/libpius/test_state.py
@@ -9,7 +9,7 @@ from libpius.state import SignState
 
 class TestFileConversion(unittest.TestCase):
   kSAMPLE = {
-      'one': {'OUTBOUND': 'SIGNED', 'INBOUND': None},
+      'one': {'OUTBOUNDD': 'SIGNED', 'INBOUND': None},
       'two': {'OUTBOUND': 'SIGNED', 'INBOUND': None},
       'three': {'OUTBOUND': 'SIGNED', 'INBOUND': None},
   }
```

**Comment by aristocrates** [comment]
I ran pius installed from this ebuild from my local overlay to sign some keys and it ran without error for me on python 3.7

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{7,8} )
+
+inherit distutils-r1
+
+DESCRIPTION="A tool for signing and email all UIDs on a set of PGP keys"
+HOMEPAGE="https://www.phildev.net/pius/"
```
Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild:11]: Could you also add the Github page as 2ndary HOMEPAGE here?

File: app-crypt/pius/pius-3.0.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+IUSE="test"
+RESTRICT="!test? ( test )"
+
+DEPEND=">=app-crypt/gnupg-2.0.0"
+RDEPEND="${DEPEND}
+	dev-lang/perl"
+
+src_test() {
+	"${S}"/run_tests || die "Tests failed"
+}
```
Comment (juippis) [inline app-crypt/pius/pius-3.0.0.ebuild:26]: You'll want to use `python_test` here so it runs the test phase through on every python implementation installed and enabled. Now it only uses your default one. Check the content of *run_tests* file, it looks like `python3 -m unittest discover -s test/* -v` where you can just replace `python3` with `${EPYTHON}`. However this test looks so simple you can most likely get it to work by calling `distutils_enable_test unittests` from *distutils-r1.eclass* that you use.

Comment (aristocrates) [reply app-crypt/pius/pius-3.0.0.ebuild]: Unfortunately it looks like `distutils_enable_tests unittests` doesn't provide a way to pass in `-s [arg]` https://github.com/gentoo/gentoo/blob/990a108c6aaae27d968207b6b75da9eb5a9d0cd9/eclass/distutils-r1.eclass#L422-L428, and then `python3 -m unittest discover` finds 0 tests to run:

```
>>> Test phase: app-crypt/pius-3.0.0
 * python3_7: running distutils-r1_run_phase python_test

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
 * python3_8: running distutils-r1_run_phase python_test

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
```

`touch test/__init__.py && touch test/libpius/__init__.py` makes `python -m unittest discover` just work (https://stackoverflow.com/a/29715336/2469311), which might be worth submitting upstream.

With this `python_test`:

```bash
python_test() {
       ${EPYTHON} -m unittest discover -s test/* -v
}
```

tests are run on all enabled python versions:

```
>>> Test phase: app-crypt/pius-3.0.0
 * python3_7: running distutils-r1_run_phase python_test
test_convert_from_v1 (test_state.TestFileConversion) ... ok
test_convert_from_v2 (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_new_data (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_no_new_data (test_state.TestFileConversion) ... ok
test_read_v1_writes_v3 (test_state.TestFileConversion) ... ok
test_read_v2_writes_v3 (test_state.TestFileConversion) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.013s

OK
 * python3_8: running distutils-r1_run_phase python_test
test_convert_from_v1 (test_state.TestFileConversion) ... ok
test_convert_from_v2 (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_new_data (test_state.TestFileConversion) ... ok
test_read_empty_writes_v3_with_no_new_data (test_state.TestFileConversion) ... ok
test_read_v1_writes_v3 (test_state.TestFileConversion) ... ok
test_read_v2_writes_v3 (test_state.TestFileConversion) ... ok

----------------------------------------------------------------------
Ran 6 tests in 0.019s

OK
```

Also note that the unit tests themselves all pass on python 3.6, which should probably also be reported upstream.

**Comment by jaymzh** [comment]
yeah my unittests are hardly comprehensive. They only cover the code converting between various versions of the config file format, because (1) it's easy to break back-compat and (2) that stuff is easy to write unittests for. They're just there to make my life easier. There has never been any attempt any any reasonable coverage (and given that the software just forks off gpg, there's not a lot that can be tested without significant mocking that would be quite cumbersome). That said, test contributions are welcome.

**Pull request closed without merging**